framework/include/base/MooseApp.h

```diff
@@ -719,18 +719,45 @@ class MooseApp : public ConsoleStreamInterface,
    */
   const MooseMesh * masterDisplacedMesh() const { return _master_displaced_mesh; }
 
+  /**
+   * Class that is used as a parameter to setFinalMeshGeneratorName()
+   *  that allows only MeshGeneratorMesh methods to call this methods
+   */
+  class SetFinalMeshGeneratorNameKey
+  {
+    friend class MeshGeneratorMesh;
+    SetFinalMeshGeneratorNameKey() {}
+    SetFinalMeshGeneratorNameKey(const SetFinalMeshGeneratorNameKey &) {}
+  };
+
   /**
    * Set final mesh generator name
+   *
+   * Guarded by the SetFinalMeshGeneartorNameKey so that it can only be called
+   * by MeshGeneratorMesh.
    */
-  void setFinalMeshGeneratorName(const std::string & generator_name);
+  void setFinalMeshGeneratorName(const std::string & generator_name,
+                                 const SetFinalMeshGeneratorNameKey);
```

Comment:
How does this work? The copy construction is considered to happen in the caller?

Reply:
Fascinating. Testing this on some small code examples, I see that the copy construction does indeed occur in the caller. I learned something new about C++ today.

Reply:
I think a comment in the doxygen about the fact that copy construction occurs in the caller would be helpful for a developer trying to understand the code. You could also pass by reference and forego a developer wondering about the copy construction; in that case you could also explicitly delete the copy constructor (and move constructor) in `SetFinalMeshGeneratorNameKey`

Reply:
I like the comment idea.

The reason I used the copy constructor is exactly this - required less code in the key class. Could also make the key class a macro (I did this in a few other objects), but I think that may be confusing to developers

Reply:
You mean the extra line of
```c++
KeyClass(KeyClass &&) = delete;
```
? I would argue that this extra line clarifies intent and does not rely on a developer knowing that if there is user-defined copy constructor, then the move constructor is not implicitly declared/defined